dotnet/src/SemanticKernel.Abstractions/AI/ChatCompletion/ChatHistory.cs

```diff
@@ -18,6 +18,14 @@ public class ChatHistory : IList<ChatMessageContent>, IReadOnlyList<ChatMessageC
     /// <summary>The messages.</summary>
     private readonly List<ChatMessageContent> _messages;
 
+    private Action<ChatMessageContent>? _overrideAdd;
```

Comment:
Can you help me understand this? Why are all of these callbacks / the derived type needed?

Reply:
TLDR, I cant have a ChatHistory : IEnumerable<ChatMessage>.

Basically any change that happens by reference in a ChatHistory I need to replicate in the `IEnumerable<ChatMessage>` when using the ChatClients, you will notice I have a private class ChatMessageHistory within the `AutoFunctionInvocationContext` that I need to keep tracking whatever happens to a `ChatHistory` and keep the `ChatMessages` updated.

This is necessary as ChatHistory can be manipulated by reference, and is how `ChatHistoryReducer` actually works.